File: elixir/rotational-cipher/rotational_cipher.exs
defmodule RotationalCipher do
  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift) do
    text
    |> to_charlist
    |> Enum.map(&rotate_letter(&1, shift))
    |> to_string
  end

  defp rotate_letter(letter, shift) when letter in ?a .. ?z, do: encode(?a, letter, shift)
  defp rotate_letter(letter, shift) when letter in ?A .. ?Z, do: encode(?A, letter, shift)
  defp rotate_letter(letter, _), do: letter

  defp encode(base, letter, shift), do: base + rem((letter - base) + shift, 26)
end

File: python/rotational-cipher/rotational_cipher.py
import string

LETTERS_IN_ALPHABET = 26

def shift(letter, key, last):
  shifted = ord(letter) + key

  if shifted > ord(last):
    shifted -= LETTERS_IN_ALPHABET

  return chr(shifted)

def encode(letter, key):
  if letter in string.ascii_lowercase:
    return shift(letter, key, 'z')

  if letter in string.ascii_uppercase:
    return shift(letter, key, 'Z')

  return letter

def rotate(plain_text, key):
  return ''.join([encode(letter, key) for letter in plain_text])
